File: leetcode/00000-00099/00054_spiral-matrix/Solution.ts
/**
 * ID:    00054
 * Title: Spiral Matrix
 * Difficulty: Medium
 * Description: Given an m x n matrix, return all elements of the matrix in spiral order.
 *
 * Example 1:
 *
 * Input: matrix = [[1,2,3],[4,5,6],[7,8,9]] Output: [1,2,3,6,9,8,7,4,5]
 *
 * Example 2:
 *
 * Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]] Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 *
 * Constraints:
 *
 * m == matrix.length
 * n == matrix[i].length
 * 1 <= m, n <= 10
 * -100 <= matrix[i][j] <= 100
 */
function spiralOrder(matrix: number[][]): number[] {
  let res = new Array()
  let x = 0
  let y = 0
  let minRow = 1
  let maxRow = matrix.length
  let minCol = 0
  let maxCol = matrix[0].length
  let total = maxCol * maxRow
  let currentDirection = 0 // 0: right, 1: down, 2: left, 3:up
  if (maxRow === 1) {
    return matrix[0]
  }
  if (maxCol === 1) {
    return matrix.reduce((pre, cur) => pre.concat(cur), [])
  }
  for (let i = 0; i < total; i++) {
    res.push(matrix[x][y])
    if (currentDirection === 0) {
      y++
      if (y === maxCol - 1) {
        currentDirection = 1
        maxCol--
      }
    }
    else if (currentDirection === 1) {
      x++
      if (x === maxRow - 1) {
        currentDirection = 2
        maxRow--
      }
    } else if (currentDirection === 2) {
      y--
      if (y === minCol) {
        currentDirection = 3
        minCol++
      }
    } else if (currentDirection === 3) {
      x--
      if (x === minRow) {
        currentDirection = 0
        minRow++
      }
    }
  }
  return res
};

function test_00054() {
  let matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
  console.log(spiralOrder(matrix));
  matrix = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
  console.log(spiralOrder(matrix));
  matrix = [[3], [2], [1]]
  console.log(spiralOrder(matrix));
  matrix = [[3, 4, 5]]
  console.log(spiralOrder(matrix));
}

test_00054()

File: leetcode/00000-00099/00054_spiral-matrix/Solution.py
#!/usr/bin/env python3
# -*- coding:UTF-8 -*-
'''
 ID:    00054
 Title: Spiral Matrix
 Difficulty: Medium
 Description: Given an m x n matrix, return all elements of the matrix in spiral order.
 *
 * Example 1:
 *
 * Input: matrix = [[1,2,3],[4,5,6],[7,8,9]] Output: [1,2,3,6,9,8,7,4,5]
 *
 * Example 2:
 *
 * Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]] Output: [1,2,3,4,8,12,11,10,9,5,6,7]
 *
 * Constraints:
 *
 * m == matrix.length
 * n == matrix[i].length
 * 1 <= m, n <= 10
 * -100 <= matrix[i][j] <= 100
'''


from curses.ascii import SO


class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        if not matrix:
            return []
        m, n = len(matrix), len(matrix[0])
        res = []
        i, j = 0, 0
        while m > 0 and n > 0:
            if m == 1:
                for k in range(n):
                    res.append(matrix[i][j + k])
                break
            if n == 1:
                for k in range(m):
                    res.append(matrix[i + k][j])
                break
            for k in range(n - 1):
                res.append(matrix[i][j + k])
            for k in range(m - 1):
                res.append(matrix[i + k][j + n - 1])
            for k in range(n - 1):
                res.append(matrix[i + m - 1][j + k])
            for k in range(m - 1):
                res.append(matrix[i + k][j])
            i += 1
            j += 1
            m -= 2
            n -= 2
        return res


def test_00054():
    print(Solution.spiralOrder(Solution, [[1, 2, 3], [4, 5, 6], [7, 8, 9]]))
    print(Solution.spiralOrder(
        Solution, [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]))
    pass


if __name__ == '__main__':
    test_00054()
